edgedb-tokio/src/tutorial.md

```diff
@@ -0,0 +1,725 @@
+[`create_client`][crate::create_client] in this case
+is encouraged.
+```
+
+# Queries with the client
+
+Here are the simplified signatures of the client methods involving queries:
+
+(Note: R here means a type that implements [QueryResult](https://docs.rs/edgedb-protocol/0.4.0/edgedb_protocol/trait.QueryResult.html))
+
+```
+fn query -> Result<Vec<R>, Error>
+fn query_json -> Result<Json, Error>
+
+fn query_single -> Result<Option<R>, Error>
+fn query_single_json -> Result<Option<Json>>
+
+fn query_required_single -> Result<R, Error>
+fn query_required_single_json -> Result<Json, Error>
+```
+
+By the way, the two `_required_single` methods just call the `_single` methods with an `.ok_or_else()` inside so structurally there is nothing different about them.
+
+These methods all take a *query* and *arguments*.
+
+A regular EdgeQL query without arguments looks like this:
+
+```
+with 
+    message1 := 'Hello there', 
+    message2 := 'General Kenobi', 
+    select message1 ++ ' ' ++ message2;
+```
+
+And the same query with arguments:
+
+```
+with 
+    message1 := <str>$0, 
+    message2 := <str>$1, 
+    select message1 ++ ' ' ++ message2;
+```
+
+In the EdgeQL REPL you are prompted to enter arguments:
+
+```
+edgedb> with
+....... message1 := <str>$0,
+....... message2 := <str>$1,
+....... select message1 ++ ' ' ++ message2;
+Parameter <str>$0: Hello there
+Parameter <str>$1: General Kenobi
+{'Hello there General Kenobi'}
+```
+
+But when using the Rust client there is no prompt to do so. Arguments also have to be in the order $0, $1, and so on as opposed to in the REPL where they can be named, as above.
+
+The `()` unit type [implements QueryArgs](https://docs.rs/edgedb-protocol/latest/edgedb_protocol/query_arg/trait.QueryArgs.html#impl-QueryArgs-for-()) and is used when no arguments are present so `&()` is a pretty common sight when using the Rust client:
+
+```rust
+client.query("This just returns a string", &());
+```
+
+These methods take two generic parameters:
+
+
+```rust
+let query_res = client.query_required_single::<String, ()>
+    ("select {'This just returns a string'}", &()).await?;
+    // or
+let query_res = client.query_required_single::<String, _>
+    ("select {'This just returns a string'}", &()).await?;
+```
+    
+Declaring the type up front tends to look neater than the turbofish syntax:
+
+```rust
+let query_res: String = client.query_required_single
+    ("select {'This just returns a string'}", &()).await?;
+```
+
+# Sample queries
+
+## When cardinality is guaranteed to be 1
+
+Using the `.query()` method works fine in any case, but returns a Vec of results. In this case we return a `Result<Vec<String>>`:
+
+```rust
+let query = "select {'This just returns a string'}";
+let query_res: Result<Vec<String>, Error> = client.query(query, &()).await;
+```
+
+If you know that only a single result will be returned, using .query_required_single() or .query_single() will be more ergonomic:
```

Comment:
In this case it's because it returns a Vec<String> (like an array of strings) even if there is just one string inside, making you have to index at [0] (quick but not good Rust practice) or use .get(0) to safely pull it out which requires a bit of handling in case nothing is there at index 0. Also it's just a bit awkward to get a Vec when you know there's only one item being returned.

I changed the example a bit to show the post-handling signature (adding a ? in Rust automatically handles the error for you which is really cool).